route/grafananet.go

```diff
@@ -143,130 +147,116 @@ func NewGrafanaNet(key, prefix, sub, regex, addr, apiKey, schemasFile string, sp
 		Min:    100 * time.Millisecond,
 		Max:    time.Minute,
 		Factor: 1.5,
 		Jitter: true,
 	}
-	body := new(bytes.Buffer)
-	for data := range route.writeQueues[shard] {
-		for {
-			pre := time.Now()
-			body.Reset()
-			snappyBody := snappy.NewWriter(body)
-			snappyBody.Write(data)
-			snappyBody.Close()
-			req, err := http.NewRequest("POST", route.addr, body)
-			if err != nil {
-				panic(err)
-			}
-			req.Header.Add("Authorization", "Bearer "+route.apiKey)
-			req.Header.Add("Content-Type", "rt-metric-binary-snappy")
-			resp, err := route.client.Do(req)
-			diff := time.Since(pre)
-			if err == nil && resp.StatusCode >= 200 && resp.StatusCode < 300 {
-				b.Reset()
-				log.Info("GrafanaNet sent metrics in %s -msg size %d", diff, len(data))
-
-				route.tickFlushSize.Update(int64(len(data)))
-				route.durationTickFlush.Update(diff)
-
-				ioutil.ReadAll(resp.Body)
-				resp.Body.Close()
-				break
-			}
-			route.numErrFlush.Inc(1)
-			dur := b.Duration()
-			if err != nil {
-				log.Warning("GrafanaNet failed to submit data: %s will try again in %s (this attempt took %s)", err, dur, diff)
-			} else {
-				buf := make([]byte, 300)
-				n, _ := resp.Body.Read(buf)
-				log.Warning("GrafanaNet failed to submit data: http %d - %s will try again in %s (this attempt took %s)", resp.StatusCode, buf[:n], dur, diff)
-
-				ioutil.ReadAll(resp.Body)
-				resp.Body.Close()
-			}
+	var dur time.Duration
+	for dur, err = route.flush(req); err != nil; {
+		route.numErrFlush.Inc(1)
+		b := boff.Duration()
+		log.Warning("GrafanaNet failed to submit data: %s - will try again in %s (this attempt took %s)", err.Error(), b, dur)
+		time.Sleep(b)
+	}
+	log.Info("GrafanaNet sent metrics in %s -msg size %d", dur, len(metrics))
+	route.durationTickFlush.Update(dur)
+	route.tickFlushSize.Update(int64(len(metrics)))
+	return metrics[:0]
+}
 
-			time.Sleep(dur)
-		}
+func (route *GrafanaNet) flush(req *http.Request) (time.Duration, error) {
+	pre := time.Now()
+	resp, err := route.client.Do(req)
+	dur := time.Since(pre)
+	if err == nil && resp.StatusCode >= 200 && resp.StatusCode < 300 {
+		ioutil.ReadAll(resp.Body)
+		resp.Body.Close()
+		return dur, nil
 	}
-	route.wg.Done()
+	if err != nil {
+		return dur, err
+	}
+	buf := make([]byte, 300)
+	n, _ := resp.Body.Read(buf)
+	ioutil.ReadAll(resp.Body)
+	resp.Body.Close()
+	return dur, fmt.Errorf("http %d - %s", resp.StatusCode, buf[:n])
 }
 
+// Dispatch takes in the requested buf or drops it if blocking mode and queue of the shard is full
 func (route *GrafanaNet) Dispatch(buf []byte) {
-	//conf := route.config.Load().(Config)
 	// should return as quickly as possible
 	log.Info("route %s sending to dest %s: %s", route.key, route.addr, buf)
-	route.dispatch(route.buf, buf, route.numBuffered, route.numDropBuffFull)
+	buf = bytes.TrimSpace(buf)
+	index := bytes.Index(buf, []byte(" "))
+	if index == -1 {
+		log.Error("RouteGrafanaNet: invalid message")
+		return
+	}
+
+	key := buf[:index]
+	route.hasher.Reset()
```

Comment:
Is route.Dispatch() ever called concurrently?  if so, we cant use a shared hasher like this.

Reply:
yes. will fix